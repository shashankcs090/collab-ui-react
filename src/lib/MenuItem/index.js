import React from 'react';
import PropTypes from 'prop-types';
import { EventOverlay, ListItem, Icon } from '@collab-ui/react/';

export default class MenuItem extends React.Component {
  static displayName = 'MenuItem';

  state = {
    anchorRef: null,
  };

  handleKeyDown = e => {
    const { handleKeyDown } = this.context;
    const { index } = this.props;
    handleKeyDown && handleKeyDown(index, this, e);
  };
  
  handleClick = e => {
    const { handleClick } = this.context;
    const { index } = this.props;
    handleClick && handleClick(index, this, e);
  };
  
  render () {
    const {
      isOpen,
      focus,
      isHeader,
      disabled,
      value,
      className,
      ariaLabel,
      subMenu,
      selectedValue,
      separator,
    } = this.props;
    return (
      <div
        className={
          'cui-menu-item' +
          `${(className && ` ${className}`) || ''}`
        }
        aria-expanded={isOpen}
        aria-haspopup={!!subMenu}
      >
        <ListItem
          onClick={this.handleClick}
          onKeyDown={this.handleKeyDown}
          ref={ref => !this.state.anchorRef && this.setState({anchorRef: ref})}
          focus={focus}
          disabled={disabled}
          isReadOnly={isHeader}
          active={isOpen}
          value={value}
          role="menuitem"
          focusOnLoad
          separator={separator}
        >
          <div className="cui-menu-item__content">
            {this.props.children}
          </div>
          <div className="cui-menu-item__selected-value" title={selectedValue}>
            {subMenu && selectedValue}
          </div>
          <div className="cui-menu-item__arrow">
            {subMenu && <Icon name="arrow-right_16"/>}
          </div>
        </ListItem>
        { isOpen ?
        <EventOverlay
          anchorNode={this.state.anchorRef}
          isOpen={isOpen}
          direction="right-top"
          closeOnClick={false}
        >
          <div
            aria-label={ariaLabel}
            role="menu"
          >
            {subMenu}
          </div>
        </EventOverlay> : null }
      </div>
    );
  }
}

MenuItem.propTypes = {
  children: PropTypes.node,
  subMenu: PropTypes.array,
  isOpen: PropTypes.bool,
  focus: PropTypes.bool,
  isHeader: PropTypes.bool,
  disabled: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.object,
    PropTypes.array
  ]),
  selectedValue: PropTypes.string,
  className: PropTypes.string,
  shouldCloseMenu: PropTypes.bool,
  ariaLabel: PropTypes.string,
  index: PropTypes.array,
  menuItem: PropTypes.array,
  separator: PropTypes.bool,
};

MenuItem.contextTypes = {
  handleClick: PropTypes.func,
  handleKeyDown: PropTypes.func,
};

MenuItem.defaultProps = {
  children: null,
  subMenu: null,
  isOpen: false,
  focus: false,
  isHeader: false,
  disabled: false,
  value: '',
  className: '',
  shouldCloseMenu: false,
  ariaLabel: '',
  selectedValue: '',
  separator: false,
};
