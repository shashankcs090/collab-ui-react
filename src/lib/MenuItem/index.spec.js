import React from 'react';
import { shallow, mount } from 'enzyme';
import MenuItem from '../MenuItem';

describe('tests for <MenuItem />', () => {
  it('should render a MenuItem', () => {
    const wrapper = shallow(
      <MenuItem>one</MenuItem>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('when MenuItem is open it should display the subMenu', () => {
    const wrapper = shallow(
      <MenuItem subMenu={[<span key="0">SubMenu</span>]} isOpen>one</MenuItem>
    );
    expect(wrapper.props()['aria-expanded']).toEqual(true);
    expect(wrapper.props()['aria-haspopup']).toEqual(true);
    expect(wrapper.props().children[0].props.active).toEqual(true);

    expect(wrapper.props().children[1].props.isOpen).toEqual(true);
    expect(wrapper.props().children[1].props.direction).toEqual("right-top");
    expect(wrapper.props().children[1].props.closeOnClick).toEqual(false);
    expect(wrapper.find('Icon').exists()).toEqual(true);

  });

  it('when MenuItem is not open it should not display the subMenu', () => {
    const wrapper = shallow(
      <MenuItem subMenu={[<span key="0">SubMenu</span>]}>one</MenuItem>
    );
    expect(wrapper.props()['aria-expanded']).toEqual(false);
    expect(wrapper.props()['aria-haspopup']).toEqual(true);
    expect(wrapper.props().children[0].props.active).toEqual(false);
    expect(wrapper.props().children[1]).toEqual(null);
  });

  it('should display the selectedValue of the menu if selectedValue prop is set', () => {
    const wrapper = mount(
      <MenuItem selectedValue="SubMenu" subMenu={[<span key="0">SubMenu</span>]}>one</MenuItem>
    );
    expect(wrapper.find('.cui-menu-item__selected-value').text()).toEqual("SubMenu");
  });

  it('should call handleClick function of context when ListItem is clicked', () => {
    const onClick = jest.fn();
    const wrapper = mount(
      <MenuItem onClick={onClick} subMenu={[<span key="0">SubMenu</span>]}>one</MenuItem>,
        { context: { handleClick: onClick } }
    );
    const listItem = wrapper.find('.cui-list-item');
    listItem.simulate('click');
    expect(onClick).toHaveBeenCalled();

  });

  it('should call handleKeyDown function of context when keyDown is fired on ListItem', () => {
    const onKeyDown = jest.fn();
    const wrapper = mount(
      <MenuItem onKeyDown={onKeyDown} subMenu={[<span key="0">SubMenu</span>]}>one</MenuItem>,
        { context: { handleKeyDown: onKeyDown } }
    );
    const listItem = wrapper.find('.cui-list-item');
    listItem.simulate('keydown');
    expect(onKeyDown).toHaveBeenCalled();
  });


});
