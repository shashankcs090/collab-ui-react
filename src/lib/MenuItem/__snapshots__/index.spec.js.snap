// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <MenuItem /> should render a MenuItem 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuItem
    ariaLabel=""
    className=""
    disabled={false}
    focus={false}
    isHeader={false}
    isOpen={false}
    selectedValue=""
    separator={false}
    shouldCloseMenu={false}
    subMenu={null}
    value=""
  >
    one
  </MenuItem>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "aria-expanded": false,
      "aria-haspopup": false,
      "children": Array [
        <ListItem
          active={false}
          className=""
          customAnchorNode={null}
          customRefProp=""
          disabled={false}
          focus={false}
          focusOnLoad={true}
          id={null}
          isReadOnly={false}
          itemIndex={null}
          label=""
          link=""
          onClick={[Function]}
          onKeyDown={[Function]}
          refName="navLink"
          role="menuitem"
          separator={false}
          type=""
          value=""
        >
          <div
            className="cui-menu-item__content"
          >
            one
          </div>
          <div
            className="cui-menu-item__selected-value"
            title=""
          />
          <div
            className="cui-menu-item__arrow"
          />
        </ListItem>,
        null,
      ],
      "className": "cui-menu-item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": false,
          "children": Array [
            <div
              className="cui-menu-item__content"
            >
              one
            </div>,
            <div
              className="cui-menu-item__selected-value"
              title=""
            />,
            <div
              className="cui-menu-item__arrow"
            />,
          ],
          "className": "",
          "customAnchorNode": null,
          "customRefProp": "",
          "disabled": false,
          "focus": false,
          "focusOnLoad": true,
          "id": null,
          "isReadOnly": false,
          "itemIndex": null,
          "label": "",
          "link": "",
          "onClick": [Function],
          "onKeyDown": [Function],
          "refName": "navLink",
          "role": "menuitem",
          "separator": false,
          "type": "",
          "value": "",
        },
        "ref": [Function],
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "one",
              "className": "cui-menu-item__content",
            },
            "ref": null,
            "rendered": "one",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": null,
              "className": "cui-menu-item__selected-value",
              "title": "",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": null,
              "className": "cui-menu-item__arrow",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": [Function],
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-expanded": false,
        "aria-haspopup": false,
        "children": Array [
          <ListItem
            active={false}
            className=""
            customAnchorNode={null}
            customRefProp=""
            disabled={false}
            focus={false}
            focusOnLoad={true}
            id={null}
            isReadOnly={false}
            itemIndex={null}
            label=""
            link=""
            onClick={[Function]}
            onKeyDown={[Function]}
            refName="navLink"
            role="menuitem"
            separator={false}
            type=""
            value=""
          >
            <div
              className="cui-menu-item__content"
            >
              one
            </div>
            <div
              className="cui-menu-item__selected-value"
              title=""
            />
            <div
              className="cui-menu-item__arrow"
            />
          </ListItem>,
          null,
        ],
        "className": "cui-menu-item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "children": Array [
              <div
                className="cui-menu-item__content"
              >
                one
              </div>,
              <div
                className="cui-menu-item__selected-value"
                title=""
              />,
              <div
                className="cui-menu-item__arrow"
              />,
            ],
            "className": "",
            "customAnchorNode": null,
            "customRefProp": "",
            "disabled": false,
            "focus": false,
            "focusOnLoad": true,
            "id": null,
            "isReadOnly": false,
            "itemIndex": null,
            "label": "",
            "link": "",
            "onClick": [Function],
            "onKeyDown": [Function],
            "refName": "navLink",
            "role": "menuitem",
            "separator": false,
            "type": "",
            "value": "",
          },
          "ref": [Function],
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "one",
                "className": "cui-menu-item__content",
              },
              "ref": null,
              "rendered": "one",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": null,
                "className": "cui-menu-item__selected-value",
                "title": "",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": null,
                "className": "cui-menu-item__arrow",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": [Function],
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
